Source/WebCore/Modules/notifications/NotificationJSONParser.cpp

```diff
@@ -111,41 +71,62 @@ static std::optional<unsigned long long> parseAppBadgeString(const String& badge
     });
 }
 
-ExceptionOr<NotificationPayload> NotificationJSONParser::parseNotificationPayload(const JSON::Object& object)
+ExceptionOr<NotificationPayload> NotificationJSONParser::parseNotificationPayload(const JSON::Object& outerObject)
 {
-    URL defaultActionURL;
-    if (auto value = object.getValue(defaultActionURLKey())) {
+    std::optional<int> declarativeWebPushValue;
+    if (auto value = outerObject.getValue(declarativeWebPushKey)) {
+        if (value->type() == JSON::Value::Type::Integer || value->type() == JSON::Value::Type::Double)
+            declarativeWebPushValue = value->asInteger();
+    }
+    if (!declarativeWebPushValue)
+        return Exception { ExceptionCode::NotFoundError, makeString("Push message does not have a valid '"_s, declarativeWebPushKey, "' integer value"_s) };
+    if (*declarativeWebPushValue != expectedDeclarativeWebPushValue)
+        return Exception { ExceptionCode::NotSupportedError, makeString("Push message does not have expected '"_s, declarativeWebPushKey, "' value"_s) };
+
+    auto notificationObject = outerObject.getValue(notificationKey);
+    if (!notificationObject)
+        return Exception { ExceptionCode::SyntaxError, makeString("Push message does not have '"_s, notificationKey, "' value"_s) };
+
+    if (notificationObject->type() != JSON::Value::Type::Object)
+        return Exception { ExceptionCode::SyntaxError, makeString("Push message '"_s, notificationKey, "' value is not the expected object type"_s) };
+
+    auto protectedObject = notificationObject->asObject();
+    RELEASE_ASSERT(protectedObject);
+    auto& object = *protectedObject;
+
+    URL navigationURL;
+    if (auto value = object.getValue(navigationURLKey)) {
         if (value->type() != JSON::Value::Type::String)
-            return Exception { ExceptionCode::SyntaxError, makeString("Push message with Notification disposition: '"_s, defaultActionURLKey(), "' member is specified but is not a string"_s) };
+            return Exception { ExceptionCode::SyntaxError, makeString("Push message with Notification disposition: '"_s, navigationURLKey, "' member is specified but is not a string"_s) };
 
-        defaultActionURL = URL { value->asString() };
+        navigationURL = URL { value->asString() };
```

Comment:
The specification uses the scope URL as the base URL.

Reply:
> I'm also not seeing the changes here to the Notification object and to PushEvent (we're using the push event now and not pushnotification. I think we should have those changes before we enable by default.

These were missed due to not having an easy way to look at the current spec changes.
They're trivial to implement, but move this patch beyond "small changes since reviewed"

So I'll land this and followup.